tls: Add ability to encrypt TLS key on disk

orbitalturtle opened the pull request:
This PR replaces #4476 . Since I've been making some of the changes for @gkrizek and it's been a bit tricky to manage making PRs into his PRs... we thought it would be nice to have a fresh start. :)

Depends on #6573

- Rebased this on top of #6500. The first six commits are redundant... If #6500 goes through first, then this will be just a few commits
- The other major change in this PR is refactoring the getTLSConfig method

Graham's original description:

Currently LND writes the TLS private key to disk in plaintext. If you are running in an untrusted environment this could be a problem as someone could use the key to snoop on your API traffic and steal credentials. This pull request creates a way to encrypt the TLS key on disk so even if your server is compromised or others have access, they can't spy on your traffic.

In these changes, we add a new flag to lnd called --tlsencryptkey. When this is enabled, lnd encrypts the TLS key using the wallet's seed and writes the encrypted blob to disk. Because the key is encrypted to the wallet's seed, that means we can only use the TLS pair when the wallet is unlocked. This would leave the WalletUnlocker service without TLS. To fix this problem, we create a temporary TLS pair specifically for the WalletUnlocker RPC. This temporary TLS key is only stored in memory and the temporary TLS cert is written to disk at the same path as tlscertpath, but with .tmp appended to the end of the filename. After unlocking, lnd deletes the cert from disk and clears the private key from memory. Then uses the main TLS certificate for all other RPCs.

Of course, this changes the flow of unlocking/init a little bit. If you use --tlsencryptkey on lnd, then you'll need a way to access the temporary TLS cert for verification. The simplest way to do this is to use lncli on the same host lnd is running on and point --tlscertpath at the temporary cert. If that's not possible then the user will need to come up with a way to copy the temporary certificate out of the lnd server and get it to their client device.

Summary of Changes:

- Moves the crypto.go file from chanbackup into its own package called lnencrypt. This was needed because we can't directly reference the encryptPayloadToWriter and decryptPayloadFromReader functions from the chanbackup package due to circular dependency issues. Also, I think it's very likely that future enhancements will also want to use these same functions. Thus I've moved them to their own package. (cherry-picked from https://github.com/lightningnetwork/lnd/pull/4458)

- Creates a new flag (--tlsencryptkey) that tells lnd to encrypt the TLS private key on disk. If set, it will generate temporary TLS certs for the WalletUnlocker RPC. Allows users to prevent spying on API traffic in untrusted environments.

- Add necessary tests and documentation

orbitalturtle commented:
Thanks @bitromortac! Yes this PR depends on this one: https://github.com/lightningnetwork/lnd/pull/6573, so might be worth giving that one a review first. 

(Also noticed both PRs need a rebase. Will do that sometime today.)

lightninglabs-deploy commented:
@sputn1ck: review reminder
@bitromortac: review reminder

lightninglabs-deploy commented:
@sputn1ck: review reminder
@bitromortac: review reminder

lightninglabs-deploy commented:
@sputn1ck: review reminder
@bitromortac: review reminder

lightninglabs-deploy commented:
@sputn1ck: review reminder
@bitromortac: review reminder

lightninglabs-deploy commented:
@sputn1ck: review reminder
@bitromortac: review reminder

lightninglabs-deploy commented:
@bitromortac: review reminder

lightninglabs-deploy commented:
@sputn1ck: review reminder
@bitromortac: review reminder

guggero commented on lnd.go:
```diff
@@ -630,105 +668,394 @@ func Main(cfg *Config, lisCfg ListenerCfg, implCfg *ImplementationCfg,
+				return err
+			}
+
+			if t.cfg.TLSEncryptKey {
```
As the linter also complains, this is a three level nested `if` and will be very hard to understand by anyone looking at it. Can you please refactor this part before we put any review work into this?

orbitalturtle replied:
Yes, makes sense, absolutely. Spent some time refactoring that.

lightninglabs-deploy commented:
@sputn1ck: review reminder
@bitromortac: review reminder
@orbitalturtle, remember to re-request review from reviewers when ready

guggero reviewed (changes requested):
Thanks for the update! The refactor is a step into the right direction, but not quite enough IMHO.
It is still very hard to reason about all the different code paths that can be taken with all the different combinations of features or states (encryption on vs. off, Let's Encrypt on vs. off, cert exists vs. needs to be generated, cert is expired vs. ok).
I left a few suggestions in the comments but I think encapsulating the TLS manager more and approaching it more from what the caller (the main `lnd` server) needs is probably the way to go. For example having a method `LoadEphemeralCert()`, then later calling `LoadLongtermCert()` to switch over to the actual cert. The caller shouldn't need to worry about what fields need to be set on the struct (so things that are just needed for the specific call should be parameters while persistent values should be set on creation).

Also, I'm not sure in general how this feature can be used in a secure way. If an emphemeral cert is created each time we start up, how can I distinguish that from an attacker's ephemeral cert? I would need a way to get the certificate file to verify it by copying it from the file system. Otherwise I risk sending my wallet unlock password to an unauthenticated service. I'm curious, how are you addressing that issue in your setup?

gkrizek commented:
> Also, I'm not sure in general how this feature can be used in a secure way.

I can address this specifically because this is something I thought a lot about several years ago when I first made this PR. We handle this by delivering the ephemeral TLS cert out of band (through our site) for effective validation. Of course in our case the user has to sort of trust that we are giving them the right cert, but that's acceptable in our model. Generally speaking, the cert will always have to be delivered/validated out of band from LND.

I thought about adding in functionality like Webhook notifications or uploading the ephemeral TLS cert to a location into LND, but determined that was out of scope for LND. If a user is leveraging this then it's up to them to validate the ephemeral TLS cert just like they have to do with the persistent TLS cert. So I don't think it degrades security at all, but it does add some additional complexity for the user that will need to be solved for. (can easily be done with automation)

guggero commented:
> 

Right, I now remember some of the discussions around webhooks and uploading the cert somewhere.
Maybe some of the security tradeoffs and considerations (especially the "out of band" part) could be a bit more explicit in the doc. Currently it just mentions "If you aren't able to run `lncli` on the host `lnd` is running on, then you'll need to copy the temporary certificate from the host onto whatever device you're using." but doesn't really go into why it's important to verify the ephemeral cert specifically.

orbitalturtle commented on lnd.go:
```diff
@@ -542,6 +562,21 @@ func Main(cfg *Config, lisCfg ListenerCfg, implCfg *ImplementationCfg,
+	// real KeyRing
+	if cfg.TLSEncryptKey {
+		// Update tlsManager with actual key.
+		tlsManager.keyRing = activeChainControl.KeyRing
```
Good call

orbitalturtle commented on lnd.go:
```diff
@@ -865,6 +1109,50 @@ func (t *TLSManager) setUpLetsEncrypt(tlsCfg *tls.Config) func() {
 	return cleanUp
 }
 
+// reloadCertificate deletes the temporary certificate file and generates a new
```
the key is just passed around in memory, but the cert file is saved as "tls.cert.tmp"

orbitalturtle commented:
@guggero Thanks for the review Oli! Yeah your feedback makes a lot of sense. In my initial refactor I was probably too conservative in trying not to break how things as they currently work, but makes sense that we needed to push the refactoring and clarification further. I pushed up a much deeper refactor, which hopefully makes the code much easier to read. LMK what you think & if this matches sort of what you were talking about in regards to a state machine?
 
I plan to spend some time running some more tests to make sure it's working as well as before (since there are so many different paths it could take), but let me know if this refactor is looking better.

lightninglabs-deploy commented:
@guggero: review reminder
@bitromortac: review reminder

guggero commented on lnd.go:
```diff
@@ -630,105 +633,203 @@ func Main(cfg *Config, lisCfg ListenerCfg, implCfg *ImplementationCfg,
+// certMaintenance checks if the certificate IP and domains matches the config,
+// and renews the certificate if either this data is outdated or the
+// certificate is expired.
+func (t *TLSManager) certMaintenance(parsedCert *x509.Certificate,
```
nit: wrap before `parsedCert` instead to avoid this weirdly indented opening parenthesis below.

guggero commented on lnd.go:
```diff
@@ -630,105 +633,203 @@ func Main(cfg *Config, lisCfg ListenerCfg, implCfg *ImplementationCfg,
+func (t *TLSManager) createTLSPair() error {
 	// Ensure we create TLS key and certificate if they don't exist.
-	if !fileExists(cfg.TLSCertPath) && !fileExists(cfg.TLSKeyPath) {
+	if !fileExists(t.cfg.TLSCertPath) && !fileExists(t.cfg.TLSKeyPath) {
```
Invert this if so we can just return `nil` if either file exists and have the generation logic non-indented?

orbitalturtle replied:
After playing with this, I think we should keep this part the same, because a couple commits later we add more logic to this function, which seems to make this change not make as much sense imo

orbitalturtle replied:
Oh wait. Nevermind, factoring in another comment you made below, you're approach looks like the right move :+1: :zap:

guggero commented on lnd.go:
```diff
@@ -630,105 +633,203 @@ func Main(cfg *Config, lisCfg ListenerCfg, implCfg *ImplementationCfg,
 			"generating a new one")
 
-		err := os.Remove(cfg.TLSCertPath)
+		err := os.Remove(t.cfg.TLSCertPath)
```
Same in this `if`, now that we're in a method, we can just return early if there's nothing to do and have the refresh logic non-indented.

guggero commented on lnd.go:
```diff
@@ -630,105 +633,203 @@ func Main(cfg *Config, lisCfg ListenerCfg, implCfg *ImplementationCfg,
 	// the certificates.
 	cleanUp := func() {}
-	if cfg.LetsEncryptDomain != "" {
+	if t.cfg.LetsEncryptDomain != "" {
```
Again, just return early if the condition doesn't apply.

guggero commented on server_test.go:
```diff
@@ -71,7 +71,10 @@ func TestTLSAutoRegeneration(t *testing.T) {
 		RPCListeners:    rpcListeners,
 	}
-	_, _, _, cleanUp, err := getTLSConfig(cfg)
+	tlsManager := TLSManager{
```
nit: use `NewTLSManager()`.

guggero commented on lnd.go:
```diff
@@ -213,15 +213,42 @@ func Main(cfg *Config, lisCfg ListenerCfg, implCfg *ImplementationCfg,
 		return mkErr("error initializing DBs: %v", err)
 	}
 
+	var serverOpts []grpc.ServerOption
```
nit: can be a single `var` statement.

guggero commented on lnd.go:
```diff
@@ -538,6 +565,21 @@ func Main(cfg *Config, lisCfg ListenerCfg, implCfg *ImplementationCfg,
+		if err != nil {
+			err := fmt.Errorf("unable to reload TLS credentials: "+
+				"%v", err)
+			ltndLog.Error(err)
```
Since we're just logging the error, can be `ltndLog.Errorf("unable to reload ... %v", err)` directly.

guggero commented on tls_manager.go:
```diff
@@ -33,20 +61,52 @@ func NewTLSManager(cfg *Config) *TLSManager {
+	tlsCfg := cert.TLSConfFromCert(certData)
+	tlsCfg.GetCertificate = t.tlsReloader.GetCertificateFunc()
+
+	var certPath string
```
nit: avoid `else` by doing

```
certPath := t.cfg.TLSCertPath
if t.ephemeralCertPath != "" {
    certPath = t.ephemeralCertPath
}
```

guggero commented on tls_manager.go:
```diff
@@ -123,9 +185,77 @@ func (t *TLSManager) createTLSPair() error {
+			keyBytes = b.Bytes()
+		}
+
+		err = cert.WriteCertPair(
```
Since we now wrote the encrypted key and cert, can't we return early now? Otherwise we try to re-encrypt it below with `t.encryptPlaintextKey()`.

orbitalturtle replied:
Good call

guggero commented on lnd.go:
```diff
@@ -213,15 +213,42 @@ func Main(cfg *Config, lisCfg ListenerCfg, implCfg *ImplementationCfg,
+			return mkErr("unable to load TLS credentials: %v", err)
+		}
+	} else {
+		_, cleanUp, err := tlsManager.generateOrRenewCert()
```
The `getConfig` and `generateOrRenewCert` methods should be exported so we could eventually even move the TLS manager to a different package.

guggero commented on tls_manager.go:
```diff
@@ -103,18 +164,19 @@ func (t *TLSManager) generateOrRenewCert() (*tls.Config, func(), error) {
 // If a TLS pair doesn't exist yet, create them and write them to disk.
-func (t *TLSManager) createTLSPair() error {
+func (t *TLSManager) createAndWriteTLSPair(keyRing keychain.SecretKeyRing,
+) error {
```
nit: break before parameter instead.

guggero commented on tls_manager.go:
```diff
@@ -0,0 +1,272 @@
+
+// fileExists reports whether the named file or directory exists.
+// This function is taken from https://github.com/btcsuite/btcd
+func fileExists(name string) bool {
```
Can just use `lnrpc.FileExists` instead, since we have the same function in 4 different places otherwise.

orbitalturtle replied:
Sounds good. Since this function was used in other files in the main lnd package, I moved this change into a separate, small commit for more clarity.

guggero reviewed (commented):
Thanks a lot for the recent changes and refactor. That definitely helped a lot with readability and following the different code flows. Added a few suggestions to improve even more, but this is getting quite close now!

bitromortac reviewed (commented):
Trying to understand the PR a bit more currently (in-depth review follows).

I tested the flow, do I understand correctly that it would be:
- user sets up a wallet (no `--tlsencryptkey`): `tls.cert` + `tls.key` are generated
- node is restarted (with `--tlsencryptkey`): `tls.cert.tmp` is created
- user unlocks wallet using `tls.cert.tmp`
- node deletes `tls.cert.tmp`, creates `tls.key.encrypted`, switches over to `tls.cert`

Would it be possible to initialize a wallet already with `--tlsencryptkey`? Also, I think it could be helpful to give the encrypted key a different name, perhaps `tls.key.encrypted` (or something similar).

gkrizek commented:
@bitromortac Yes, you understand that flow right. It is possible to start a wallet with `--tlsencryptkey` but I don't think it should be the default if that's what you're asking.

I think renaming the key to something different could cause confusion in my opinion. having different or multiple tls.key files depending on settings seems like it'd be hard to follow.

bitromortac commented:
> Yes, you understand that flow right. It is possible to start a wallet with --tlsencryptkey but I don't think it should be the default if that's what you're asking.

Thanks @gkrizek.
Currently we error for an empty `.lnd` with `[ERR] LTND: Shutting down because error in main method: unable to load TLS credentials: open /home/user/.lnd/tls.cert: no such file or directory`, so that would need to be created first before trying to encrypt it.
Also I'm wondering for which applications we would need an unencrypted `tls.key` and why it's not a good idea to do this by default (and ensure backward compatibility with unlocking).

gkrizek commented:
@bitromortac 
> Currently we error for an empty .lnd with [ERR] LTND: Shutting down because error in main method: unable to load TLS credentials: open /home/user/.lnd/tls.cert: no such file or directory, so that would need to be created first before trying to encrypt it.

I'm not sure I understand what you're saying. Can you elaborate here?

> Also I'm wondering for which applications we would need an unencrypted tls.key and why it's not a good idea to do this by default (and ensure backward compatibility with unlocking).

I don't think there's really a problem or place where having an encrypted TLS key is bad, however the flow to unlock the node (with the temporary certificates) is a pretty different process than what has existed in LND so far. So I think forcing this new process for unlocking nodes would have negative cascading affects.

Thanks!

orbitalturtle commented:
Addressed those comments @guggero! Just one question: Looks like the linter wants me to add a timeout to the lets encrypt codes. Would it make sense to use one of the other timeouts in the config.go or should I set a new one?

And uh, I don't know why the GitHub UI hates me, but every time I go to request a review from anyone, it for some reason removes other developers' requests for review (???). So sorry for removing your request for review @sputn1ck.

guggero commented:
> Looks like the linter wants me to add a timeout to the lets encrypt codes. Would it make sense to use one of the other timeouts in the config.go or should I set a new one?

I'd just add a `var` at the top of the file for this case and set it to 5 seconds or something like that. The linter is new so it didn't catch existing instances of `http.Server` that we should probably also address. I'm going to add an issue for that so we can fix it everywhere (REST server, profiling server, let's encrypt).

bitromortac commented on server.go:
```diff
@@ -1640,8 +1645,24 @@ func (s *server) createLivenessMonitor(cfg *Config, cc *chainreg.ChainControl) {
 		func() error {
-			_, parsedCert, err := cert.LoadCert(
-				cfg.TLSCertPath, cfg.TLSKeyPath,
+			certBytes, keyBytes, err := cert.GetCertBytesFromPath(
```
I think we could hide this in `TLSManager` as well, it would place the complexity into the more specialized struct.

orbitalturtle replied:
Good call

bitromortac commented on tls_manager.go:
```diff
@@ -0,0 +1,517 @@
+// options needed for gRPC and REST.
+type TLSManager struct {
+	cfg         *Config
+	tlsReloader *cert.TLSReloader
```
Would be great if you could document `tlsReloader` as switching mechanics of certificates is still not 100% clear to me.

orbitalturtle commented on lnd.go:
```diff
@@ -539,6 +567,20 @@ func Main(cfg *Config, lisCfg ListenerCfg, implCfg *ImplementationCfg,
+	// TLSConfig. Now we want to remove that and load the persistent
+	// TLSConfig. The wallet is unlocked at this point so we can use the
+	// real KeyRing.
+	if cfg.TLSEncryptKey {
```
Got it, so pretty much encapsulate everything in this `if` statement (including the if statemnet)? And then expose SetCertificateAfterUnlock(KeyRing) instead of LoadPermanentCertificate?

gkrizek commented:
@bitromortac maybe @guggero can help us decide the default behavior. Although this is more secure, I do not think it should be the default behavior because it changes the way nodes are unlocked significantly. I think this new and different way of unlocking would be bad for existing LND users. The preference would be to allow them to upgrade when they are ready.

>  Would it be too involved to keep tls.cert in sync with the currently active certificate? That way we would keep backward compatibility with unlocking. 

I think this is true only if the user is using lncli from the same instance that LND is running on. If they are communicating with the node remotely it wouldn't provide any benefits. You could still need to grab the cert to validate the request. Therefore, I think overwriting the `tls.cert` file back and forth doesn't make a lot of sense and could cause confusion.

orbitalturtle commented on tls_manager.go:
```diff
@@ -0,0 +1,517 @@
+// it encrypts the TLS key. It also returns the certificate configuration
+// options needed for gRPC and REST.
+type TLSManager struct {
+	cfg         *Config
```
Like this?

orbitalturtle commented:
> > I'm not sure I understand what you're saying. Can you elaborate here?
> 
> @gkrizek Yes, sorry for not being more explicit. The code as in this PR, running lnd for the first time with the `--tlsencryptkey` flag is not compatible with an uninitialized `.lnd` home folder, as the flow expects a `tls.cert/key`pair in the first call of `GetConfig` to exist and errors with a file not found error. Is that intentional or a deeper reason behind it?

Oops yup looks like that was a bug. Nice catch. I also addressed all your other comments @bitromortac, when you have a chance to take a look.

guggero commented:
> @bitromortac maybe @guggero can help us decide the default behavior. Although this is more secure, I do not think it should be the default behavior because it changes the way nodes are unlocked significantly.

I agree with Graham here, we shouldn't make this the default because it will make unlocking through `lncli` quite different because another cert needs to be used.

bitromortac commented:
> I think this is true only if the user is using lncli from the same instance that LND is running on. If they are communicating with the node remotely it wouldn't provide any benefits. You could still need to grab the cert to validate the request. Therefore, I think overwriting the tls.cert file back and forth doesn't make a lot of sense and could cause confusion.

> I agree with Graham here, we shouldn't make this the default because it will make unlocking through `lncli` quite different because another cert needs to be used.

Yes, that makes. I didn't really consider that scenario, usually when I'm restarting my node I unlock it from the same machine, but there is a use case where the node hoster is different from the manager :+1: (which is also part of the motivation of this PR).

bitromortac reviewed (changes requested):
Nice, almost done from my side, I left some suggestions for improvement. I also tested that the file not found bug is absent :+1:.

lightninglabs-deploy commented:
@orbitalturtle, remember to re-request review from reviewers when ready

orbitalturtle commented on tls_manager.go:
```diff
@@ -0,0 +1,650 @@
+				"TLS credentials: %v", err)
+		}
+	} else {
+		_, cleanUp, err := t.GenerateOrRenewCert()
```
Good call, added a function in createAndWriteTLSPair that takes care of this.

I also added a small test block to ensure this gets caught.

orbitalturtle commented:
@bitromortac Sounds great, made those changes :+1: @guggero & @sputn1ck lmk if there's anything else I can do to improve the PR. I'd re-request a review above, but having trouble with the UI lately.

lightninglabs-deploy commented:
@orbitalturtle, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@sputn1ck: review reminder
@orbitalturtle, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@sputn1ck: review reminder
@orbitalturtle, remember to re-request review from reviewers when ready

bitromortac commented on tls_manager.go:
```diff
@@ -265,6 +466,145 @@ func (t *TLSManager) setUpLetsEncrypt(certData *tls.Certificate,
+		return nil, err
+	}
+	t.setEphemeralSettings(keyBytes, certBytes, t.cfg.TLSCertPath+".tmp")
+	t.ephemeralKey = keyBytes
```
I think those lines are not needed anymore.

bitromortac reviewed (commented):
Nice coverage of the different flows, test are looking good :+1:. Mostly nits from my side, I also tested for the error after running a second time without the flag.

lightninglabs-deploy commented:
@orbitalturtle, remember to re-request review from reviewers when ready

orbitalturtle commented:
Thanks for the reviews @bitromortac and @sputn1ck. @bitromortac I made those changes ~ re: a couple of your comments, I think I had done something funny when rebasing, but should be fixed now :+1:

sputn1ck reviewed (approved):
Great PR! 

Did different local tests again, all went as expected

Code LGTM

bitromortac reviewed (approved):
LGTM :rocket:, thank you for the PR! There are some minor comments left.

orbitalturtle commented on tls_manager_test.go:
```diff
@@ -0,0 +1,371 @@
+
+// TestTLSEncryptDiscrepancies tests a couple of things:
+// 1) If we have cfg.TLSEncryptKey set, but the tls file saved to disk is not
+// encrypted, GenerateOrRenewCert encrypts the file and rewrites it to disk.
```
Cool, made these changes. I like the names you suggested. The only one I was iffy on was changing `loadEphemeralCertificate` to `generateEphemeralCertPair`, because I like how the "load" corresponds to `LoadPermanentCertificate`, which does a similar thing. But I changed the rest of them.

orbitalturtle commented:
Fabulous, @bitromortac I made those changes, thanks again for the reviews :)

guggero reviewed (commented):
Thanks for the quick updates! The new naming looks good.

A final set of nits then we're good to go :tada:

orbitalturtle commented:
Great made those changes @guggero
